pydis_site/apps/api/signals.py

```diff
@@ -0,0 +1,12 @@
+from django.db.models.signals import pre_delete
+from django.dispatch import receiver
+
+from pydis_site.apps.api.models.bot import Role, User
+
+
+@receiver(signal=pre_delete, sender=Role)
+def delete_role_from_user(sender: Role, instance: Role, **kwargs) -> None:
+    """Unassigns the Role (instance) that is being deleted from every user that has it."""
+    for user in User.objects.filter(roles__contains=[instance.id]):
+        del user.roles[user.roles.index(instance.id)]
+        user.save()
```

Comment:
Is there not a way to do this in batch? We should let SQL do this

Reply:
Oh, I see your concern. That wasn't a wise choice indeed. Although, I think there's a way to solve this using Django ORM.

Reply:
It seems like, the Django ORM, F objects in particular does not support ArrayFields. 
However, I've never really worked with arrayfields using raw SQL.

Reply:
This isn't possible with plain PostgreSQL, except if we were to use triggers, and I would currently stay away from that personally.

The original approach here was the correct way: foreign key constraints. But since this got removed due to performance issues (waiting on reply from a Django expert friend on this topic), this is a fine approach.